### DISCUSSION TITLE
strange tear lines when adding dots to on top of squares in keyboard-canvas-response

### DISCUSSION THREAD

**Discussion by Max-Lovell** [jspsych/jsPsych#1953] [Q&A] [answered]
can anyone elaborate as to why this happens?

![image](https://user-images.githubusercontent.com/45423050/123983956-2b643a80-d9bc-11eb-9d67-b62e53ccf6c9.png)

```<!DOCTYPE html>
<html>
    <head>
        <title>Experimental task</title>
        <script src="jspsych-6.3.1/jspsych.js"></script>
        <script src="jspsych-6.3.1/plugins/jspsych-canvas-keyboard-response.js"></script>
        <link href="jspsych-6.3.1/css/jspsych.css" rel="stylesheet" type="text/css"></link>
        <link href="style.css" rel="stylesheet"></link>
    </head>
    <body>
    </body>
    <script>

    var timeline = [];
   var squareWidth = 250;

function drawStim(c){
    var ctx = c.getContext('2d')
    var l_x = 75
    var r_x = 475
    var co_y = 145
    ctx.fillRect(l_x, co_y, squareWidth, squareWidth);
    ctx.fillRect(r_x, co_y, squareWidth, squareWidth);

    var cellSize = 10;
    var arr = [...Array(625).keys()];
    var dotIndex = jsPsych.randomization.shuffle(arr);

    var mat_no = 0;
     for (var x = l_x ; x < l_x + squareWidth; x += cellSize){
         for (var y = co_y ; y < co_y + squareWidth; y += cellSize){
             if (dotIndex[mat_no] < (312 + 70)){
                 ctx.arc(x + (cellSize/2) , y + (cellSize/2), 2, 0, Math.PI*2);
                 ctx.fillStyle = "white";
                 ctx.fill();
             }
     mat_no++;
     }
}}

    var trial = {
        type: 'canvas-keyboard-response',
        canvas_size: [540, 800],
        stimulus: drawStim,
        choices: ['e','i'],
        prompt: `<p style="position:relative; top:150px; font-size:28px">right or left?</p>`,
        data: {shape: 'rectangle'}
    };

    timeline.push(trial);

    jsPsych.init({
      timeline: timeline
    });
    </script>
</html>

```````
CSS file (can be removed):

#jspsych-canvas-stimulus {
	position: absolute;
	top: 0;
	left: 0;
	right: 0;
	margin:  auto;
}

#jspsych-content.text {
	position: absolute;
	top: 0;
	left: 0;
	right: 0;
	margin:  auto;
}

**Comment by becky-gilbert** [comment, marked as answer]
I think what's happening is that each drawing path is always starting from the last x/y coordinate, then moving to the x/y coordinate given in `ctx.arc`, then drawing the circle and filling it in. This works fine in the first row of dots because the x value is the same and the y value keeps moving down, so the drawing path never creates a fillable shape other than the dot. But when the x value increases to the next row and the y value goes back down, the drawing path crosses itself, and then `ctx.fill()` fills in everything in the joined shape.

Anyway the fix is easy: just try adding `ctx.beginPath();` right before the `ctx.arc` line to re-start the drawing path each time.

**Reply by becky-gilbert** [reply]
By the way, in case you're curious, I figured this out by drawing just a small subset of dots and printing out the coordinates, and then increasing the number of dots and looking to see when the tear lines started occurring. 
![canvas-dots](https://user-images.githubusercontent.com/9041788/124026590-ba665800-d9a6-11eb-97b1-dd464d28f3cc.PNG)
![canvas-dots-tear](https://user-images.githubusercontent.com/9041788/124026602-be927580-d9a6-11eb-9ae8-48feb27d562a.PNG)

**Reply by Max-Lovell** [reply]
perfect! thanks very much for the explanation, fix and debugging tip becky!
